File: server/src/services/shares.ts
import { Scanner } from "./scanner"
import { ThumbnailGenerator } from "./thumbnails"
import { MediaCache } from "./cache"
import type { ShareScanResult, MediaStats, MediaFile } from "../types"
import { MEDIA_SHARES } from "../config"

export class MediaShare {
  private scanner: Scanner
  private thumbnailGenerator: ThumbnailGenerator
  private cache: MediaCache
  private processedFiles = new Set<string>()

  constructor() {
    this.scanner = new Scanner(
      (file) => this.handleFileFound(file),
      (filePath) => this.handleFileRemoved(filePath),
      (dirPath, shareName) => this.handleDirectoryFound(dirPath, shareName)
    )
    this.thumbnailGenerator = new ThumbnailGenerator()
    this.cache = new MediaCache()
  }

  async init() {
    await this.cache.load()
    await this.scanner.startWatching()

    for (const shareName of Object.keys(MEDIA_SHARES)) {
      setImmediate(() => this.backgroundScan(shareName))
    }
  }

  async getShareFiles(
    shareName: string,
    subPath = ""
  ): Promise<ShareScanResult> {
    if (!(shareName in MEDIA_SHARES)) {
      throw new Error(`Share ${shareName} not found`)
    }

    const { files, directories } = this.cache.getShareFiles(shareName, subPath)

    return {
      files,
      directories,
      isScanning: false,
    }
  }

  getStats(): MediaStats {
    const cacheStats = this.cache.getStats()
    return {
      shares: Object.keys(MEDIA_SHARES).length,
      totalFiles: cacheStats.totalFiles,
      totalDirectories: cacheStats.totalDirectories,
      thumbnailQueueSize: 0,
      backgroundWorkers: 0,
      watchers: 1,
    }
  }

  findTrackById(fileId: string) {
    return this.cache.findTrackById(fileId)
  }

  getThumbnailPath(thumbId: string) {
    return this.thumbnailGenerator.getThumbnailPath(thumbId)
  }

  private async handleFileFound(file: MediaFile) {
    if (this.processedFiles.has(file.id)) {
      return
    }

    this.processedFiles.add(file.id)

    console.log(`[MediaShare] File found: ${file.path}`)

    this.cache.addOrUpdateTrack(file)
    await this.cache.save()
    this.thumbnailGenerator.queueThumbnail(file)
  }

  private async handleFileRemoved(filePath: string) {
    console.log(`[MediaShare] File removed: ${filePath}`)
  }

  private async handleDirectoryFound(dirPath: string, shareName: string) {
    console.log(`[MediaShare] Directory found: ${dirPath} in ${shareName}`)
    this.cache.addDirectory(dirPath, shareName)
    await this.cache.save()
  }

  private async backgroundScan(shareName: string) {
    try {
      console.log(`[MediaShare] Starting background scan for ${shareName}`)
      const result = await this.scanner.scanShare(shareName)

      let processedCount = 0
      for (const file of result.files) {
        await this.handleFileFound(file)
        processedCount++

        if (processedCount % 10 === 0) {
          console.log(
            `[MediaShare] Background scan for share "${shareName}" processed ${processedCount}/${result.files.length} files`
          )
        }
      }
      console.log(
        `[MediaShare] Background scan for share "${shareName}" completed. Processed ${processedCount} files`
      )
    } catch (error) {
      console.error(
        `[MediaShare] Error during background scan for ${shareName}: ${error}`
      )
    }
  }

  async shutdown() {
    await this.scanner.stop()
    await this.thumbnailGenerator.shutdown()
    await this.cache.save()
  }
}

File: mpv-remote-server/services/shares.py
import asyncio
from typing import Set

from services.scanner import Scanner
from services.thumbnails import ThumbnailGenerator
from services.cache import MediaCache
from models.model import ShareScanResult, MediaStats, MediaFile
from config import settings


class MediaShare:
    def __init__(self):
        self.scanner = Scanner(
            self.handle_file_found,
            self.handle_file_removed,
            self.handle_directory_found,
        )
        self.thumbnail_generator = ThumbnailGenerator()
        self.cache = MediaCache()
        self.processed_files: Set[str] = set()

    async def init(self):
        await self.cache.load()
        await self.thumbnail_generator.start()
        await self.scanner.start_watching()

        for share_name in settings.media_shares.keys():
            asyncio.create_task(self.background_scan(share_name))

    async def get_share_files(
        self, share_name: str, sub_path: str = ""
    ) -> ShareScanResult:
        if share_name not in settings.media_shares:
            raise ValueError(f"Share {share_name} not found")

        files, directories = self.cache.get_share_files(share_name, sub_path)

        return ShareScanResult(files=files, directories=directories, isScanning=False)

    def get_stats(self) -> MediaStats:
        total_files, total_directories = self.cache.get_stats()
        return MediaStats(
            shares=len(settings.media_shares),
            totalFiles=total_files,
            totalDirectories=total_directories,
            thumbnailQueueSize=self.thumbnail_generator.queue_size,
            backgroundWorkers=0,
            watchers=1,
        )

    def find_track_by_id(self, file_id: str):
        return self.cache.find_track_by_id(file_id)

    def get_thumbnail_path(self, thumb_id: str):
        return self.thumbnail_generator.get_thumbnail_path(thumb_id)

    async def handle_file_found(self, file: MediaFile):
        if file.id in self.processed_files:
            return

        self.processed_files.add(file.id)

        print(f"[MediaShare] File found: {file.path}")

        self.cache.add_or_update_track(file)
        await self.cache.save()
        self.thumbnail_generator.queue_thumbnail(file)

    async def handle_file_removed(self, file_path: str):
        print(f"[MediaShare] File removed: {file_path}")

    async def handle_directory_found(self, dir_path: str, share_name: str):
        print(f"[MediaShare] Directory found: {dir_path} in {share_name}")
        self.cache.add_directory(dir_path, share_name)
        await self.cache.save()

    async def background_scan(self, share_name: str):
        try:
            print(f"[MediaShare] Starting background scan for {share_name}")
            result = await self.scanner.scan_share(share_name)

            processed_count = 0
            for file in result.files:
                await self.handle_file_found(file)
                processed_count += 1

                if processed_count % 10 == 0:
                    print(
                        f"[MediaShare] Background scan for share '{share_name}' processed {processed_count}/{len(result.files)} files"
                    )

            print(
                f"[MediaShare] Background scan for share '{share_name}' completed. Processed {processed_count} files"
            )
        except Exception as error:
            print(
                f"[MediaShare] Error during background scan for {share_name}: {error}"
            )

    async def shutdown(self):
        await self.scanner.stop()
        await self.thumbnail_generator.shutdown()
        await self.cache.save()
